src/ImageSharp/Common/Helpers/SimdUtils.Avx2Intrinsics.cs

```diff
@@ -91,6 +91,190 @@ public static class Avx2Intrinsics
+                VerifySpanInput(channel2, dest, Vector256<byte>.Count);
+
+                int n = dest.Length / Vector256<byte>.Count;
+
+                ref Vector256<byte> source0Base =
+                    ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(channel0));
+                ref Vector256<byte> source1Base =
+                    ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(channel1));
+                ref Vector256<byte> source2Base =
+                    ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(channel2));
+
+                ref Vector256<byte> destBase =
+                    ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(dest));
+
+                Vector256<byte> s0Mask0 = Vector256.Create(0, -1, -1, 1, -1, -1, 2, -1, -1, 3, -1, -1, 4, -1, -1, 5, -1, -1, 6, -1, -1, 7, -1, -1, 8, -1, -1, 9, -1, -1, 10, -1).AsByte();
+                Vector256<byte> s0Mask1 = Vector256.Create(-1, 11, -1, -1, 12, -1, -1, 13, -1, -1, 14, -1, -1, 15, -1, -1, 0, -1, -1, 1, -1, -1, 2, -1, -1, 3, -1, -1, 4, -1, -1, 5).AsByte();
+                Vector256<byte> s0Mask2 = Vector256.Create(-1, -1, 6, -1, -1, 7, -1, -1, 8, -1, -1, 9, -1, -1, 10, -1, -1, 11, -1, -1, 12, -1, -1, 13, -1, -1, 14, -1, -1, 15, -1, -1).AsByte();
+
+                Vector256<byte> s1Mask0 = Vector256.Create(-1, 0, -1, -1, 1, -1, -1, 2, -1, -1, 3, -1, -1, 4, -1, -1, 5, -1, -1, 6, -1, -1, 7, -1, -1, 8, -1, -1, 9, -1, -1, 10).AsByte();
+                Vector256<byte> s1Mask1 = Vector256.Create(-1, -1, 11, -1, -1, 12, -1, -1, 13, -1, -1, 14, -1, -1, 15, -1, -1, 0, -1, -1, 1, -1, -1, 2, -1, -1, 3, -1, -1, 4, -1, -1).AsByte();
+                Vector256<byte> s1Mask2 = Vector256.Create(5, -1, -1, 6, -1, -1, 7, -1, -1, 8, -1, -1, 9, -1, -1, 10, -1, -1, 11, -1, -1, 12, -1, -1, 13, -1, -1, 14, -1, -1, 15, -1).AsByte();
+
+                Vector256<byte> s2Mask0 = Vector256.Create(-1, -1, 0, -1, -1, 1, -1, -1, 2, -1, -1, 3, -1, -1, 4, -1, -1, 5, -1, -1, 6, -1, -1, 7, -1, -1, 8, -1, -1, 9, -1, -1).AsByte();
+                Vector256<byte> s2Mask1 = Vector256.Create(10, -1, -1, 11, -1, -1, 12, -1, -1, 13, -1, -1, 14, -1, -1, 15, -1, -1, 0, -1, -1, 1, -1, -1, 2, -1, -1, 3, -1, -1, 4, -1).AsByte();
+                Vector256<byte> s2Mask2 = Vector256.Create(-1, 5, -1, -1, 6, -1, -1, 7, -1, -1, 8, -1, -1, 9, -1, -1, 10, -1, -1, 11, -1, -1, 12, -1, -1, 13, -1, -1, 14, -1, -1, 15).AsByte();
+
+                for (int i = 0, j = 0; j < n; i += 1, j += 3)
+                {
+                    Vector256<byte> s0 = Unsafe.Add(ref source0Base, i);
+                    Vector256<byte> s1 = Unsafe.Add(ref source1Base, i);
+                    Vector256<byte> s2 = Unsafe.Add(ref source2Base, i);
+
+                    Vector256<byte> loS0 = Avx2.Permute2x128(s0, s0, 0);
+                    Vector256<byte> loS1 = Avx2.Permute2x128(s1, s1, 0);
+                    Vector256<byte> loS2 = Avx2.Permute2x128(s2, s2, 0);
```

Comment:
Again, all these permutes and shuffles will be quite expensive.  What I do for these 3 channel cases is treat them the same as 4 channel right up until the end (in this case with a dummy zero vector), and then do a single shuffle to pack the 12 good bytes out of each 16 together.  That does mean overrunning the end of your buffer by 8 bytes for the AVX2 implementation, though, so you'd have to adjust your remainder/cleanup length by 8 to compensate.  It's worth doing some benchmarks to see if the difference if you're not clear on it.

Reply:
That single shuffle... How would that work? I can't figure it out without multiple permutes.

Reply:
So, assuming this one starts off the same as the 4-channel version, you'll end up with vectors that look like this:

```
BGRx:BGRx:BGRx:BGRx||BGRx:BGRx:BGRx:BGRx
```

With a shuffle, you can pack the BGR/RGB triplets together, putting the filler/alpha values at the end of each lane.  Then with a permute, you can cram the 12-byte valid sections of each lane together, leaving all the dummy values in the last 8 bytes.  Like such (forgive my poor ascii art skills):

```
BGRx:BGRx:BGRx:BGRx||BGRx:BGRx:BGRx:BGRx
vpshufb (0,1,2,4,5,6,8,9,a,c,d,e,3,7,b,f)
                   ||
                   \/
BGRB:GRBG:RBGR:xxxx||BGRB:GRBG:RBGR:xxxx
vpermd (0,1,2,4,5,6,3,7)
                   ||
                   \/
BGRB:GRBG:RBGR:BGRB||GRBG:RBGR:xxxx:xxxx
```

At that point, you can either write the full 32-byte vector out (but then only advance your output ref/ptr by 2 bytes), or you can write the lower lane and then `movq` the first half of the upper lane, depending on whether your at the end of your out buffer or not.

Reply:
BTW, because this one requires 3 permutes to start and then 4 to end, it may be one where SSE2 is faster than AVX2.  Would be worth benchmarking.

Reply:
Yay! I couldn't figure out the first part. Thanks!

On another note, `BroadcastVector128ToVector256` should absolutely have an overload accepting a `Vector128<T>`.  

![image](https://user-images.githubusercontent.com/385879/97120504-4e449d80-170f-11eb-9fd0-eea67b0de491.png)

Reply:
Nice!

The `vbroadcasti128` instruction only accepts a memory operand, so to use it with `Vector128<T>`, it would have to spill to memory.  You can, however, use `Vector256<T>.Create()` with the same `Vector128<T>` for both args, and it will emit the correct permute for you.